File: packages/sdk-mcp/src/exporter.ts
import type { IngestPayload, SpanInput } from '@lookspan/types';

export interface SpanExporter {
  send(spans: SpanInput[]): Promise<void>;
  flush(): Promise<void>;
}

export interface HttpExporterOptions {
  endpoint: string;
  flushIntervalMs?: number;
  maxBatchSize?: number;
  source?: string;
  fetchImpl?: typeof fetch;
  /** Extra POST attempts after the first on a transient failure. Default: 2. */
  maxRetries?: number;
  /** Base backoff between attempts; doubles each retry. Default: 200ms. */
  retryBackoffMs?: number;
}

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

export class HttpSpanExporter implements SpanExporter {
  private buffer: SpanInput[] = [];
  private timer: NodeJS.Timeout | null = null;
  private readonly flushIntervalMs: number;
  private readonly maxBatchSize: number;
  private readonly fetchImpl: typeof fetch;
  private readonly maxRetries: number;
  private readonly retryBackoffMs: number;

  constructor(private readonly options: HttpExporterOptions) {
    this.flushIntervalMs = options.flushIntervalMs ?? 1000;
    this.maxBatchSize = options.maxBatchSize ?? 100;
    this.fetchImpl = options.fetchImpl ?? fetch;
    this.maxRetries = Math.max(0, options.maxRetries ?? 2);
    this.retryBackoffMs = Math.max(0, options.retryBackoffMs ?? 200);
  }

  async send(spans: SpanInput[]): Promise<void> {
    this.buffer.push(...spans);
    if (this.buffer.length >= this.maxBatchSize) {
      await this.flush();
    } else {
      this.scheduleFlush();
    }
  }

  async flush(): Promise<void> {
    if (this.buffer.length === 0) return;
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    const batch = this.buffer;
    this.buffer = [];

    const payload: IngestPayload = {
      spans: batch,
      source: this.options.source ?? '@lookspan/mcp',
      sentAt: new Date().toISOString(),
    };
    const body = JSON.stringify(payload);

    // Telemetry must never crash or block the host app, but a single transient
    // failure (collector restarting, brief network blip) shouldn't silently
    // drop a whole batch either. Retry transient errors with exponential
    // backoff; give up on client errors (4xx) since the payload won't change.
    let lastError = '';
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0) await sleep(this.retryBackoffMs * 2 ** (attempt - 1));
      try {
        const res = await this.fetchImpl(this.options.endpoint, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body,
        });
        if (res.ok) return;
        lastError = `HTTP ${res.status}`;
        if (res.status >= 400 && res.status < 500) break; // not retryable
      } catch (err) {
        lastError = (err as Error).message;
      }
    }
    console.warn(
      `[lookspan/mcp] export failed after ${this.maxRetries + 1} attempt(s), dropped ${batch.length} span(s): ${lastError}`,
    );
  }

  private scheduleFlush(): void {
    if (this.timer) return;
    this.timer = setTimeout(() => {
      this.timer = null;
      void this.flush();
    }, this.flushIntervalMs);
  }
}

File: python/lookspan-core/src/lookspan/exporter.py
"""Span exporters — pluggable transports for sending spans to a Lookspan collector."""

from __future__ import annotations

import threading
from datetime import datetime, timezone
from typing import Protocol

import httpx

from lookspan.types import IngestPayload, Span


class SpanExporter(Protocol):
    def send(self, spans: list[Span]) -> None: ...
    def flush(self) -> None: ...
    def close(self) -> None: ...


class HttpSpanExporter:
    """Buffered HTTP exporter. Spans are batched and flushed asynchronously."""

    def __init__(
        self,
        endpoint: str,
        *,
        source: str = "lookspan-python",
        flush_interval: float = 1.0,
        max_batch_size: int = 100,
        timeout: float = 5.0,
        client: httpx.Client | None = None,
    ) -> None:
        self._endpoint = endpoint
        self._source = source
        self._flush_interval = flush_interval
        self._max_batch_size = max_batch_size
        self._buffer: list[Span] = []
        self._lock = threading.Lock()
        self._client = client or httpx.Client(timeout=timeout)
        self._timer: threading.Timer | None = None
        self._closed = False

    def send(self, spans: list[Span]) -> None:
        if self._closed:
            return
        with self._lock:
            self._buffer.extend(spans)
            if len(self._buffer) >= self._max_batch_size:
                self._flush_locked()
            else:
                self._schedule_flush()

    def flush(self) -> None:
        with self._lock:
            self._flush_locked()

    def close(self) -> None:
        self.flush()
        self._closed = True
        if self._timer:
            self._timer.cancel()
            self._timer = None
        self._client.close()

    def _schedule_flush(self) -> None:
        if self._timer is not None:
            return
        self._timer = threading.Timer(self._flush_interval, self._timer_flush)
        self._timer.daemon = True
        self._timer.start()

    def _timer_flush(self) -> None:
        with self._lock:
            self._timer = None
            self._flush_locked()

    def _flush_locked(self) -> None:
        if not self._buffer:
            return
        batch = self._buffer
        self._buffer = []
        if self._timer:
            self._timer.cancel()
            self._timer = None
        payload = IngestPayload(
            spans=batch,
            source=self._source,
            sent_at=datetime.now(tz=timezone.utc).isoformat().replace("+00:00", "Z"),
        )
        try:
            self._client.post(self._endpoint, json=payload.to_payload())
        except httpx.HTTPError as err:
            # Don't crash the user's app on telemetry failures.
            print(f"[lookspan] export failed: {err}")


__all__ = ["HttpSpanExporter", "SpanExporter"]
